src/pexprs-assertAllApplicationsAreValid.js

```diff
@@ -12,49 +12,62 @@ var pexprs = require('./pexprs');
 // Operations
 // --------------------------------------------------------------------
 
-var lexifyCount;
+var inLexifiedStackContext = [];
 
 pexprs.PExpr.prototype.assertAllApplicationsAreValid = function(ruleName, grammar) {
-  lexifyCount = 0;
+  inLexifiedStackContext = [];
+  inLexifiedStackContext.push(!common.isSyntactic(ruleName));
   this._assertAllApplicationsAreValid(ruleName, grammar);
 };
 
 pexprs.PExpr.prototype._assertAllApplicationsAreValid = common.abstract(
-  '_assertAllApplicationsAreValid'
+    '_assertAllApplicationsAreValid'
 );
 
 pexprs.any._assertAllApplicationsAreValid =
-pexprs.end._assertAllApplicationsAreValid =
-pexprs.Terminal.prototype._assertAllApplicationsAreValid =
-pexprs.Range.prototype._assertAllApplicationsAreValid =
-pexprs.Param.prototype._assertAllApplicationsAreValid =
-pexprs.UnicodeChar.prototype._assertAllApplicationsAreValid = function(ruleName, grammar) {
-  // no-op
+    pexprs.end._assertAllApplicationsAreValid =
+        pexprs.Terminal.prototype._assertAllApplicationsAreValid =
+            pexprs.Range.prototype._assertAllApplicationsAreValid =
+                pexprs.Param.prototype._assertAllApplicationsAreValid =
+                    pexprs.UnicodeChar.prototype._assertAllApplicationsAreValid = function(ruleName, grammar) {
+                      // no-op
+                    };
+
+pexprs.Syn.prototype._assertAllApplicationsAreValid = function(ruleName, grammar) {
+  inLexifiedStackContext.push(false);
+  this.expr._assertAllApplicationsAreValid(ruleName, grammar);
+  inLexifiedStackContext.pop();
 };
 
 pexprs.Lex.prototype._assertAllApplicationsAreValid = function(ruleName, grammar) {
-  lexifyCount++;
+  inLexifiedStackContext.push(true);
   this.expr._assertAllApplicationsAreValid(ruleName, grammar);
-  lexifyCount--;
+  inLexifiedStackContext.pop();
 };
 
 pexprs.Alt.prototype._assertAllApplicationsAreValid = function(ruleName, grammar) {
   for (var idx = 0; idx < this.terms.length; idx++) {
+    inLexifiedStackContext.push(inLexifiedStackContext[inLexifiedStackContext.length-1]);
```

Comment:
I don't understand why you need to add this here. Only Lex or Syn can change the context, so shouldn't they be the only ones pushing and popping?